pkg/cmd/bazci/bazci.go

```diff
@@ -66,6 +77,185 @@ tests in Teamcity as painless as possible.`,
 	}
 )
 
+type monitorBuildServer struct {
+	action           string // "build" or "test".
+	namedSetsOfFiles map[string][]builtArtifact
+	testResults      map[string][]fullTestResult
+	builtTargets     map[string]*bes.TargetComplete
+	// Send a bool value to this channel when it's time to tear down the
+	// server.
+	finished chan bool
+	wg       sync.WaitGroup
+}
+
+func newMonitorBuildServer(action string) *monitorBuildServer {
+	return &monitorBuildServer{
+		action:           action,
+		namedSetsOfFiles: make(map[string][]builtArtifact),
+		testResults:      make(map[string][]fullTestResult),
+		builtTargets:     make(map[string]*bes.TargetComplete),
+		finished:         make(chan bool, 1),
+	}
+}
+
+func (s *monitorBuildServer) Start() error {
+	conn, err := net.Listen("tcp", ":"+strconv.Itoa(port))
+	if err != nil {
+		return err
+	}
+	srv := grpc.NewServer()
+	build.RegisterPublishBuildEventServer(srv, s)
+	s.wg.Add(2)
+	go func() {
+		// This worker gracefully stops the server when the build is finished.
+		defer s.wg.Done()
+		<-s.finished
+		srv.GracefulStop()
+	}()
+	go func() {
+		// This worker runs the server.
+		defer s.wg.Done()
+		if err := srv.Serve(conn); err != nil {
+			panic(err)
+		}
+	}()
+	return nil
+}
+
+func (s *monitorBuildServer) Wait() {
+	s.wg.Wait()
+}
+
+// This function does nothing but we need it to implement PublishBuildEventServer.
```

Comment:
nit: please update the comment to reflect what it does now if you decide not to use `LastMessage` instead otherwise update the function to do nothing